File: python/path_traversal/file_download.py
import os
from flask import Flask, send_file, request

app = Flask(__name__)
BASE_DIR = "/var/www/files"

# VULN 1: open() with user-controlled filename - no path sanitization
def read_user_file(filename):
    path = BASE_DIR + "/" + filename
    with open(path, "r") as f:
        return f.read()


# VULN 2: Flask send_file with user-controlled path - allows ../../etc/passwd
@app.route("/download")
def download_file():
    filename = request.args.get("file")
    filepath = os.path.join(BASE_DIR, filename)
    return send_file(filepath)


# VULN 3: os.path.join without canonicalization - traversal via absolute path or ../
@app.route("/export")
def export_report():
    report_name = request.args.get("name")
    full_path = os.path.join("/reports/output", report_name)
    with open(full_path, "rb") as f:
        return f.read()

File: go/path_traversal/file_download.go
package pathtraversal

import (
	"net/http"
	"os"
	"path/filepath"
)

const baseDir = "/var/www/files"

// VULN 1: os.ReadFile with string concatenation - no path sanitization
func ReadUserFile(w http.ResponseWriter, r *http.Request) {
	filename := r.URL.Query().Get("file")
	path := baseDir + "/" + filename
	data, err := os.ReadFile(path)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	w.Write(data)
}

// VULN 2: http.ServeFile with user-controlled path - allows ../../etc/passwd
func DownloadFile(w http.ResponseWriter, r *http.Request) {
	filename := r.URL.Query().Get("file")
	fullPath := filepath.Join(baseDir, filename)
	http.ServeFile(w, r, fullPath)
}

// VULN 3: os.Open with user-controlled path - stream arbitrary files
func ExportReport(w http.ResponseWriter, r *http.Request) {
	reportName := r.URL.Query().Get("name")
	fullPath := filepath.Join("/reports/output", reportName)
	f, err := os.Open(fullPath)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	defer f.Close()
	data, _ := os.ReadFile(fullPath)
	w.Write(data)
}
